libcaf_core/caf/flow/op/sample.test.cpp

```diff
@@ -0,0 +1,355 @@
+        pub.push({2});
+        advance_time(1s);
+        run_flows();
+        dispatch_messages();
+        pub.push({3});
+        advance_time(1s);
+        run_flows();
+        dispatch_messages();
+        pub.close();
+        run_flows();
+        dispatch_messages();
+        auto expected = std::vector<int>{1, 2, 3};
+        check_eq(*outputs, expected);
+        check_eq(*err, caf::sec::runtime_error);
+      }
+    }
+  }
+  GIVEN("an observable that produces only an error") {
+    WHEN("calling .sample() on it") {
+      THEN("the observer receives the error") {
+        auto outputs = std::make_shared<std::vector<int>>();
+        auto err = std::make_shared<error>();
+        sys.spawn([outputs, err](caf::event_based_actor* self) {
+          self->make_observable()
+            .fail<int>(make_error(caf::sec::runtime_error))
+            .sample(1s)
+            .do_on_error([err](const error& what) { *err = what; })
+            .for_each([outputs](const int& xs) { outputs->emplace_back(xs); });
+        });
+        run_flows();
+        dispatch_messages();
+        check(outputs->empty());
+        check_eq(*err, caf::sec::runtime_error);
+      }
+    }
+  }
+}
+
+SCENARIO("samples dispose unexpected subscriptions") {
+  GIVEN("an initialized sample operator") {
+    WHEN("calling on_subscribe with unexpected subscriptions") {
+      THEN("the sample disposes them immediately") {
+        auto snk = flow::make_passive_observer<int>();
+        auto uut = raw_sub(make_observable().never<int>(),
+                           make_observable().never<int64_t>(),
+                           snk->as_observer());
+        auto data_sub = make_never_sub<int>(snk->as_observer());
+        auto ctrl_sub = make_never_sub<int>(snk->as_observer());
+        uut->fwd_on_subscribe(fwd_data, caf::flow::subscription{data_sub});
+        uut->fwd_on_subscribe(fwd_ctrl, caf::flow::subscription{ctrl_sub});
+        check(snk->subscribed());
+        check(!uut->disposed());
+        run_flows();
+        check(data_sub->disposed());
+        check(ctrl_sub->disposed());
+        uut->dispose();
+        run_flows();
+        check(uut->disposed());
+      }
+    }
+  }
+}
+
+SCENARIO("samples emit final items after an on_error event") {
+  GIVEN("an initialized sample operator") {
+    WHEN("calling on_error(data) on a sample without pending data") {
+      THEN("the sample forward on_error immediately") {
+        auto snk = flow::make_passive_observer<int>();
+        auto uut = raw_sub(make_observable().never<int>(),
+                           make_observable().never<int64_t>(),
+                           snk->as_observer());
+        snk->request(42);
+        run_flows();
+        uut->fwd_on_next(fwd_data, 1);
+        uut->fwd_on_next(fwd_data, 2);
+        uut->fwd_on_next(fwd_data, 3);
+        check_eq(uut->pending(), true);
+        uut->fwd_on_error(fwd_data, sec::runtime_error);
+        check_eq(snk->buf, std::vector{3});
```

Comment:
I don't think the operator should have produced a value here. The case is different for `buffer`, because it uses the timeout merely as a "maximum delay". However, `sample` is supposed to emit values at a fixed interval. There was no tick, hence there should be no value.

We can argue whether to still produce that final value or not, but the operator should wait for the next tick before shutting down.

Same for `on_complete` behavior. Basically: if the input flow stops, we still wait for the next tick. The tick might either emit a value or not (depending on whether `buf_` holds a value) and then checks if the operator needs to shut down.